core/primitives/src/shard_layout.rs

```diff
@@ -83,17 +98,198 @@ pub struct ShardLayoutV1 {
+    fn account_id_to_shard_id(&self, account_id: &AccountId) -> ShardId {
+        let mut shard_id: ShardId = 0;
+        for boundary_account in &self.boundary_accounts {
+            if account_id < boundary_account {
+                break;
+            }
+            shard_id += 1;
+        }
+        shard_id
+    }
+}
+
+/// A boundary of an account range. Can be the start, the end or an account id
+/// in between two shards. For example a shard layout with four shards would
+/// have one start boundary, three middle boundaries and one end boundary.
+/// e.g. Start, Middle("ccc"), Middle("kkk"), Middle("ppp"), End
+///
+/// The order of fields (Start, Middle, End) in the enum is important as it
+/// defines the derived ordering.
+#[derive(serde::Serialize, serde::Deserialize, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub enum AccountBoundary {
+    Start,
+    Middle(AccountId),
+    End,
+}
+
+impl fmt::Debug for AccountBoundary {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            AccountBoundary::Start => write!(f, "start"),
+            AccountBoundary::End => write!(f, "end"),
+            AccountBoundary::Middle(account_id) => write!(f, "{}", account_id.as_str()),
+        }
+    }
+}
+
+/// The account range of a single shard.
+#[derive(serde::Serialize, serde::Deserialize, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct AccountRange {
+    start: AccountBoundary,
+    end: AccountBoundary,
+}
+
+impl fmt::Debug for AccountRange {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let range = format!("{:?}..{:?}", self.start, self.end);
+        f.debug_struct("AccountRange").field("range", &range).finish()
+    }
+}
+
+impl AccountRange {
+    pub fn new_single_shard() -> Self {
+        Self { start: AccountBoundary::Start, end: AccountBoundary::End }
+    }
+
+    pub fn new_start(end: AccountId) -> Self {
+        Self { start: AccountBoundary::Start, end: AccountBoundary::Middle(end) }
+    }
+
+    pub fn new_end(start: AccountId) -> Self {
+        Self { start: AccountBoundary::Middle(start), end: AccountBoundary::End }
+    }
+
+    pub fn new_mid(start: AccountId, end: AccountId) -> Self {
+        Self { start: AccountBoundary::Middle(start), end: AccountBoundary::Middle(end) }
+    }
+
+    fn contains(&self, account_id: &AccountId) -> bool {
+        match (&self.start, &self.end) {
+            (AccountBoundary::Start, AccountBoundary::End) => true,
+            (AccountBoundary::Start, AccountBoundary::Middle(end)) => account_id < end,
+            (AccountBoundary::Middle(start), AccountBoundary::End) => account_id >= start,
+            (AccountBoundary::Middle(start), AccountBoundary::Middle(end)) => {
+                account_id >= start && account_id < end
+            }
+            (AccountBoundary::End, _) => panic!("invalid account range"),
+            (_, AccountBoundary::Start) => panic!("invalid account range"),
+        }
+    }
+}
```

Comment:
Alternative:

I remembered there are primitives `std::ops::RangeBounds` and `std::ops::Bound`. We can represent shard ranges using them. `RangeBounds::contains` looks exactly like this function, with exception that there is `Bound::Unbounded` which by definition means both start and end, depending on which side of range it represents.

There are also RangeFull, RangeTo, RangeFrom.

Reply:
Oh nice, thanks! I should have guessed there is some ready structure for this. I'll check it out tomorrow!

Reply:
Would be nice to use primitives in std if it's simple enough for our use case